server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1631,63 +1636,93 @@ void removeEntitlement(Entitlement entitlement, boolean regenModified) {
         }
 
         // post unbind actions
-        PoolHelper poolHelper = new PoolHelper(this, entitlement);
-        enforcer.postUnbind(consumer, poolHelper, entitlement);
+        for (Entitlement ent : entsToRevoke) {
+            PoolHelper poolHelper = new PoolHelper(this,  ent);
+            enforcer.postUnbind(ent.getConsumer(), poolHelper, ent);
+        }
+
+        List<Entitlement> batch = new ArrayList<Entitlement>();
+        for (int i = 0; i < entsToRevoke.size(); i++) {
+            Entitlement entitlement = entsToRevoke.get(i);
+            batch.add(entitlement);
 
-        if (regenModified) {
-            // Find all of the entitlements that modified the original entitlement,
-            // and regenerate those to remove the content sets.
-            // Lazy regeneration is ok here.
-            this.regenerateCertificatesOf(entitlementCurator.listModifying(entitlement), true);
+            // We work in batches of maximum size 1000.
+            if (i % 1000 == 0) {
+                Set<Entitlement> modifiedEnts = entitlementCurator.batchListModifying(batch);
+                this.regenerateCertificatesOf(modifiedEnts, true);
+                batch.clear();
+            }
         }
 
-        // If we are deleting a developer entitlement, be sure to delete the
-        // associated pool as well.
-        if (pool.isDevelopmentPool()) {
-            poolCurator.delete(pool);
+        if (!batch.isEmpty()) {
+            this.regenerateCertificatesOf(entitlementCurator.batchListModifying(batch), true);
         }
 
-        log.info("Revoked entitlement: {}", entitlement.getId());
+        log.info("Modifier entitlements done.");
+
+        Set<Consumer> distinctConsumers = entitlementCurator.getDistinctConsumers(entsToRevoke);
 
-        // If we don't care about updating other entitlements based on this one, we probably
-        // don't care about updating compliance either.
-        if (regenModified) {
-            // Check consumer's new compliance status and save:
+        log.info("Recomputing status for {} consumers.", distinctConsumers.size());
+        int i = 1;
+        for (Consumer consumer : distinctConsumers) {
+            if (i++ % 1000 == 0) {
+                consumerCurator.flush();
+            }
             complianceRules.getStatus(consumer);
         }
+        consumerCurator.flush();
 
-        sink.queueEvent(event);
+        log.info("All statuses recomputed.");
+
+        // for each deleted entitlement, create an event
+        for (Entitlement entitlement : entsToRevoke) {
+            Consumer consumer = entitlement.getConsumer();
+            Event event = eventFactory.entitlementDeleted(entitlement);
```

Comment:
The original code behaved the same way. 

```java
  Event event = eventFactory.entitlementDeleted(entitlement);
        if (!entitlement.isValid() &&
                entitlement.getPool().isUnmappedGuestPool() &&
                consumerCurator.getHost(consumer.getFact("virt.uuid"), consumer.getOwner()) == null) {
            event = eventFactory.entitlementExpired(entitlement);
            event.setMessageText(event.getMessageText() + ": " +
                i18n.tr("Unmapped guest entitlement expired without establishing a host/guest mapping."));
        }
```
Do you think it is a bug even in master?